src/Ship.cpp

```diff
@@ -379,50 +379,55 @@ inline int sign(T num)
 vector3d Ship::CalcAtmosphericForce() const
 {
 	// Data from ship.
-	double m_topCrossSec = GetShipType()->topCrossSection;
-	double m_sideCrossSec = GetShipType()->sideCrossSection;
-	double m_frontCrossSec = GetShipType()->frontCrossSection;
+	const auto topCrossSec = GetShipType()->topCrossSection;
+	const auto sideCrossSec = GetShipType()->sideCrossSection;
+	const auto frontCrossSec = GetShipType()->frontCrossSection;
 
 	// TODO: vary drag coefficient based on Reynolds number, specifically by
 	// atmospheric composition (viscosity) and airspeed (mach number).
-	double m_topDragCoeff = GetShipType()->topDragCoeff;
-	double m_sideDragCoeff = GetShipType()->sideDragCoeff;
-	double m_frontDragCoeff = GetShipType()->frontDragCoeff;
+	const auto topDragCoeff = GetShipType()->topDragCoeff;
+	const auto sideDragCoeff = GetShipType()->sideDragCoeff;
+	const auto frontDragCoeff = GetShipType()->frontDragCoeff;
 
-	double m_shipLiftCoeff = GetShipType()->shipLiftCoefficient;
+	const auto shipLiftCoeff = GetShipType()->shipLiftCoefficient;
 
 	// By converting the velocity into local space, we can apply the drag individually to each component.
-	auto m_localVel = GetVelocity() * GetOrient();
-	auto m_lVSqr = m_localVel.LengthSqr();
+	const auto localVel = GetVelocity() * GetOrient();
 
 	// The drag forces applied to the craft, in local space.
 	// TODO: verify dimensional accuracy and that we're not generating more drag than physically possible.
 	// TODO: use a different drag constant for each side (front, back, etc).
 	// This also handles (most of) the lift due to wing deflection.
-	vector3d fAtmosDrag = vector3d(
-		CalcAtmosphericDrag(m_lVSqr, m_sideCrossSec, m_sideDragCoeff),
-		CalcAtmosphericDrag(m_lVSqr, m_topCrossSec, m_topDragCoeff),
-		CalcAtmosphericDrag(m_lVSqr, m_frontCrossSec, m_frontDragCoeff));
 
-	// The direction vector of the velocity also serves to scale and sign the generated drag.
-	fAtmosDrag = fAtmosDrag * -m_localVel.NormalizedSafe();
+	// Get current atmosphere parameters
+	double pressure, density;
+	GetCurrentAtmosphericState(pressure, density);
+
+	// Precalculate common part of drag components calculation (rho / 2)
+	const auto rho_2 = density / 2.;
+	// Vector, which components are square of local airspeed vector components
+	const vector3d lv2(localVel * localVel);
+	const vector3d fAtmosDrag(
+		-sign(localVel.x) * rho_2 * lv2.x * sideCrossSec * sideDragCoeff,
+		-sign(localVel.y) * rho_2 * lv2.y * topCrossSec * topDragCoeff,
+		-sign(localVel.z) * rho_2 * lv2.z * frontCrossSec * frontDragCoeff);
```

Comment:
The original code (postmultiplication by unit vector) essentially did what your code here does, just in a more roundabout way. I like this form better, as it involves less virtual calls and is probably more correct.